dynamiqs/sparse.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+
+import equinox as eqx
+import jax.numpy as jnp
+from jaxtyping import Array, ArrayLike, ScalarLike
+
+from .qarray import QArray
+
+
+class SparseQArray(QArray):
+    diags: Array
+    offsets: tuple[int, ...] = eqx.field(static=True)
+    dims: tuple[int, ...] = eqx.field(static=True)
+
+    def __add__(
+        self, other: ScalarLike | ArrayLike | SparseQArray
+    ) -> Array | SparseQArray:
+        if isinstance(other, ScalarLike):
+            if other == 0:
+                return self
+            return self.to_dense() + other
+        elif isinstance(other, ArrayLike):
```

Comment:
It feels like you could just do `diags + other` here

Reply:
This is not the same operation. You would be adding a constant value only to the non-zero diagonals, which is not what you expect from the standard `__add__` on an array. Calling `x+2` adds a `2` to all matrix elements, so we return a dense matrix of a sparse one.

Reply:
Oh right I see. Then we want to set a warning because this does 2 things we probably don't want to
- convert a sparse a dense matrix
- add value to all elements (A + x) instead of (A + x. 1)